File: internal/integrations/kubernetes/actions.go
package kubernetes

import (
	"errors"
	"fmt"
	"regexp"
	"strings"

	"github.com/roee-hersh/hallpass/internal/catalog"
)

// alias is a named action that expands to a verb and resource.
// An empty resource means "take resource/group from the request".
type alias struct {
	name, desc  string
	verb        string
	resource    string
	group       string
	subresource string
}

var aliasList = []alias{
	{name: "pods.exec", desc: "run a command in a pod (create pods/exec)", verb: "create", resource: "pods", subresource: "exec"},
	{name: "pods.logs", desc: "read pod logs (get pods/log)", verb: "get", resource: "pods", subresource: "log"},
	{name: "pods.portforward", desc: "port-forward to a pod (create pods/portforward)", verb: "create", resource: "pods", subresource: "portforward"},
	{name: "pods.attach", desc: "attach to a pod (create pods/attach)", verb: "create", resource: "pods", subresource: "attach"},
	{name: "scale", desc: "scale the resource named in the request (update <resource>/scale)", verb: "update", subresource: "scale"},
	{name: "secrets.read", desc: "read a secret (get secrets)", verb: "get", resource: "secrets"},
	{name: "secrets.list", desc: "list secrets (list secrets)", verb: "list", resource: "secrets"},
	{name: "impersonate", desc: "impersonate the subject named in the request, default users (impersonate <resource>)", verb: "impersonate"},
	{name: "deployment.create", desc: "create a deployment (create deployments.apps)", verb: "create", resource: "deployments", group: "apps"},
	{name: "deployment.update", desc: "update a deployment (update deployments.apps)", verb: "update", resource: "deployments", group: "apps"},
	{name: "deployment.delete", desc: "delete a deployment (delete deployments.apps)", verb: "delete", resource: "deployments", group: "apps"},
	{name: "deployment.restart", desc: "restart a deployment (patch deployments.apps)", verb: "patch", resource: "deployments", group: "apps"},
	{name: "namespace.create", desc: "create a namespace (create namespaces)", verb: "create", resource: "namespaces"},
	{name: "namespace.delete", desc: "delete a namespace (delete namespaces)", verb: "delete", resource: "namespaces"},
	{name: "rbac.bind", desc: "bind a role (create rolebindings.rbac.authorization.k8s.io)", verb: "create", resource: "rolebindings", group: "rbac.authorization.k8s.io"},
}

var aliases = func() map[string]alias {
	m := map[string]alias{}
	for _, a := range aliasList {
		m[a.name] = a
	}
	return m
}()

var (
	verbRe      = regexp.MustCompile(`^[a-z]{1,32}$`)
	resourceRe  = regexp.MustCompile(`^[a-z0-9][a-z0-9-]{0,62}$`)
	groupRe     = regexp.MustCompile(`^[a-z0-9]([a-z0-9.-]{0,251}[a-z0-9])?$`)
	nameRe      = regexp.MustCompile(`^[A-Za-z0-9]([A-Za-z0-9._:-]{0,252}[A-Za-z0-9])?$`)
	namespaceRe = regexp.MustCompile(`^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?$`)
	pathRe      = regexp.MustCompile(`^/[A-Za-z0-9/._~*-]{0,1000}$`)
)

// rawSpec is a parsed raw:<verb>:<resource>[.<group>][/<subresource>] action.
type rawSpec struct {
	verb, resource, group, subresource string
}

// parseRaw accepts raw:<verb>:<resource>... and, for non-resource paths,
// the verb-only form raw:<verb>.
func parseRaw(name string) (rawSpec, error) {
	parts := strings.SplitN(name, ":", 3)
	if len(parts) < 2 || parts[0] != "raw" {
		return rawSpec{}, errors.New("raw action must be raw:<verb>:<resource>[.<group>][/<subresource>] or raw:<verb> for non-resource paths")
	}
	verb := parts[1]
	if !verbRe.MatchString(verb) {
		return rawSpec{}, fmt.Errorf("verb %q must be lowercase letters", verb)
	}
	if len(parts) == 2 {
		return rawSpec{verb: verb}, nil
	}
	res, group, sub, err := splitResource(parts[2])
	if err != nil {
		return rawSpec{}, err
	}
	return rawSpec{verb: verb, resource: res, group: group, subresource: sub}, nil
}

// splitResource parses "deployments.apps/scale" into resource, group and
// subresource. The group is everything after the first dot, as kubectl does.
func splitResource(s string) (resource, group, subresource string, err error) {
	if s == "" {
		return "", "", "", errors.New("resource is empty")
	}
	res, sub, _ := strings.Cut(s, "/")
	res, group, _ = strings.Cut(res, ".")
	if !resourceRe.MatchString(res) {
		return "", "", "", fmt.Errorf("resource %q must be a lowercase plural such as pods or deployments", res)
	}
	if group != "" && !groupRe.MatchString(group) {
		return "", "", "", fmt.Errorf("API group %q is not a valid group name", group)
	}
	if sub != "" && !resourceRe.MatchString(sub) {
		return "", "", "", fmt.Errorf("subresource %q is not valid", sub)
	}
	return res, group, sub, nil
}

// buildAttributes combines the action and the resource reference.
//
//	namespace:<ns>[?resource=<res>[.<group>]&name=<n>&subresource=<s>]
//	cluster[?resource=<res>[.<group>]&name=<n>&subresource=<s>]
//	nonresource:<path>
func buildAttributes(actionName string, res catalog.Resource) (attributes, error) {
	var a attributes
	var actRes, actGroup, actSub string
	if strings.HasPrefix(actionName, "raw:") {
		spec, err := parseRaw(actionName)
		if err != nil {
			return a, err
		}
		a.verb, actRes, actGroup, actSub = spec.verb, spec.resource, spec.group, spec.subresource
	} else {
		al, ok := aliases[actionName]
		if !ok {
			return a, fmt.Errorf("unknown action %q", actionName)
		}
		a.verb, actRes, actGroup, actSub = al.verb, al.resource, al.group, al.subresource
	}

	switch res.Type {
	case "nonresource":
		if !pathRe.MatchString(res.ID) {
			return a, fmt.Errorf("nonresource path %q must start with / and contain only URL path characters", res.ID)
		}
		if actRes != "" {
			return a, fmt.Errorf("action %s names a resource but the request is a non-resource path; use raw:<verb>", actionName)
		}
		a.nonResourcePath = res.ID
		return a, nil
	case "namespace":
		if !namespaceRe.MatchString(res.ID) {
			return a, fmt.Errorf("namespace %q is not a valid namespace name", res.ID)
		}
		a.namespace = res.ID
	case "cluster":
		if res.ID != "" {
			return a, errors.New("cluster resources take no id: use cluster?resource=nodes")
		}
	default:
		return a, fmt.Errorf("resource type %q; use namespace:<ns>, cluster or nonresource:<path>", res.Type)
	}

	reqRes, reqGroup, reqSub := "", "", ""
	if q := res.Q("resource"); q != "" {
		var err error
		reqRes, reqGroup, reqSub, err = splitResource(q)
		if err != nil {
			return a, err
		}
	}
	if s := res.Q("subresource"); s != "" {
		if !resourceRe.MatchString(s) {
			return a, fmt.Errorf("subresource %q is not valid", s)
		}
		reqSub = s
	}
	if n := res.Q("name"); n != "" {
		if !nameRe.MatchString(n) {
			return a, fmt.Errorf("name %q is not a valid object name", n)
		}
		a.name = n
	}
	if ns := res.Q("namespace"); ns != "" {
		return a, errors.New("put the namespace in the resource id: namespace:<ns>")
	}

	switch {
	case actRes != "" && reqRes != "" && (actRes != reqRes || actGroup != reqGroup):
		return a, fmt.Errorf("action %s targets %s but the request names %s", actionName, joinRes(actRes, actGroup), joinRes(reqRes, reqGroup))
	case actRes != "":
		a.resource, a.group = actRes, actGroup
	case reqRes != "":
		a.resource, a.group = reqRes, reqGroup
	case actionName == "impersonate":
		a.resource = "users"
	case strings.HasPrefix(actionName, "raw:") && !strings.Contains(actionName[4:], ":"):
		return a, fmt.Errorf("action %s is the non-resource form; use it with nonresource:<path>, or add ?resource=<plural>[.<group>] to the request", actionName)
	default:
		return a, fmt.Errorf("action %s needs a resource: add ?resource=<plural>[.<group>] to the request", actionName)
	}
	switch {
	case actSub != "" && reqSub != "" && actSub != reqSub:
		return a, fmt.Errorf("action %s targets subresource %s but the request names %s", actionName, actSub, reqSub)
	case actSub != "":
		a.subresource = actSub
	default:
		a.subresource = reqSub
	}
	return a, nil
}

func joinRes(res, group string) string {
	if group == "" {
		return res
	}
	return res + "." + group
}

File: hallpass-py/src/hallpass/integrations/kubernetes/actions.py
"""Named actions and the parsing of raw actions and resource references."""

from __future__ import annotations

import re
from dataclasses import dataclass

from hallpass.core.catalog import Resource
from hallpass.core.errors import go_quote

__all__ = [
    "ALIASES",
    "ALIAS_LIST",
    "Alias",
    "Attributes",
    "RawSpec",
    "build_attributes",
    "join_res",
    "parse_raw",
    "split_resource",
]


@dataclass(frozen=True)
class Alias:
    """A named action that expands to a verb and resource. An empty
    resource means "take resource/group from the request"."""

    name: str
    desc: str
    verb: str
    resource: str = ""
    group: str = ""
    subresource: str = ""


ALIAS_LIST: tuple[Alias, ...] = (
    Alias("pods.exec", "run a command in a pod (create pods/exec)", "create", resource="pods", subresource="exec"),
    Alias("pods.logs", "read pod logs (get pods/log)", "get", resource="pods", subresource="log"),
    Alias("pods.portforward", "port-forward to a pod (create pods/portforward)", "create", resource="pods", subresource="portforward"),
    Alias("pods.attach", "attach to a pod (create pods/attach)", "create", resource="pods", subresource="attach"),
    Alias("scale", "scale the resource named in the request (update <resource>/scale)", "update", subresource="scale"),
    Alias("secrets.read", "read a secret (get secrets)", "get", resource="secrets"),
    Alias("secrets.list", "list secrets (list secrets)", "list", resource="secrets"),
    Alias("impersonate", "impersonate the subject named in the request, default users (impersonate <resource>)", "impersonate"),
    Alias("deployment.create", "create a deployment (create deployments.apps)", "create", resource="deployments", group="apps"),
    Alias("deployment.update", "update a deployment (update deployments.apps)", "update", resource="deployments", group="apps"),
    Alias("deployment.delete", "delete a deployment (delete deployments.apps)", "delete", resource="deployments", group="apps"),
    Alias("deployment.restart", "restart a deployment (patch deployments.apps)", "patch", resource="deployments", group="apps"),
    Alias("namespace.create", "create a namespace (create namespaces)", "create", resource="namespaces"),
    Alias("namespace.delete", "delete a namespace (delete namespaces)", "delete", resource="namespaces"),
    Alias("rbac.bind", "bind a role (create rolebindings.rbac.authorization.k8s.io)", "create", resource="rolebindings", group="rbac.authorization.k8s.io"),
)

ALIASES: dict[str, Alias] = {a.name: a for a in ALIAS_LIST}

# Go's regexps are anchored with ^...$ and have no trailing-newline
# exception; these are always used with fullmatch.
VERB_RE = re.compile(r"[a-z]{1,32}")
RESOURCE_RE = re.compile(r"[a-z0-9][a-z0-9-]{0,62}")
GROUP_RE = re.compile(r"[a-z0-9]([a-z0-9.-]{0,251}[a-z0-9])?")
NAME_RE = re.compile(r"[A-Za-z0-9]([A-Za-z0-9._:-]{0,252}[A-Za-z0-9])?")
NAMESPACE_RE = re.compile(r"[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?")
PATH_RE = re.compile(r"/[A-Za-z0-9/._~*-]{0,1000}")


def _ok(r: re.Pattern[str], s: str) -> bool:
    return r.fullmatch(s) is not None


@dataclass(frozen=True)
class RawSpec:
    """A parsed raw:<verb>:<resource>[.<group>][/<subresource>] action."""

    verb: str
    resource: str = ""
    group: str = ""
    subresource: str = ""


def parse_raw(name: str) -> RawSpec:
    """Accept raw:<verb>:<resource>... and, for non-resource paths, the
    verb-only form raw:<verb>. Raises ValueError."""
    parts = name.split(":", 2)
    if len(parts) < 2 or parts[0] != "raw":
        raise ValueError("raw action must be raw:<verb>:<resource>[.<group>][/<subresource>] or raw:<verb> for non-resource paths")
    verb = parts[1]
    if not _ok(VERB_RE, verb):
        raise ValueError(f"verb {go_quote(verb)} must be lowercase letters")
    if len(parts) == 2:
        return RawSpec(verb=verb)
    res, group, sub = split_resource(parts[2])
    return RawSpec(verb=verb, resource=res, group=group, subresource=sub)


def split_resource(s: str) -> tuple[str, str, str]:
    """Parse "deployments.apps/scale" into resource, group and subresource.
    The group is everything after the first dot, as kubectl does."""
    if s == "":
        raise ValueError("resource is empty")
    res, _, sub = s.partition("/")
    res, _, group = res.partition(".")
    if not _ok(RESOURCE_RE, res):
        raise ValueError(f"resource {go_quote(res)} must be a lowercase plural such as pods or deployments")
    if group != "" and not _ok(GROUP_RE, group):
        raise ValueError(f"API group {go_quote(group)} is not a valid group name")
    if sub != "" and not _ok(RESOURCE_RE, sub):
        raise ValueError(f"subresource {go_quote(sub)} is not valid")
    return res, group, sub


@dataclass
class Attributes:
    """What goes into the review."""

    namespace: str = ""
    verb: str = ""
    group: str = ""
    resource: str = ""
    name: str = ""
    subresource: str = ""
    non_resource_path: str = ""


def build_attributes(action_name: str, res: Resource) -> Attributes:
    """Combine the action and the resource reference. Raises ValueError.

    namespace:<ns>[?resource=<res>[.<group>]&name=<n>&subresource=<s>]
    cluster[?resource=<res>[.<group>]&name=<n>&subresource=<s>]
    nonresource:<path>
    """
    a = Attributes()
    if action_name.startswith("raw:"):
        spec = parse_raw(action_name)
        a.verb, act_res, act_group, act_sub = spec.verb, spec.resource, spec.group, spec.subresource
    else:
        al = ALIASES.get(action_name)
        if al is None:
            raise ValueError(f"unknown action {go_quote(action_name)}")
        a.verb, act_res, act_group, act_sub = al.verb, al.resource, al.group, al.subresource

    if res.type == "nonresource":
        if not _ok(PATH_RE, res.id):
            raise ValueError(f"nonresource path {go_quote(res.id)} must start with / and contain only URL path characters")
        if act_res != "":
            raise ValueError(f"action {action_name} names a resource but the request is a non-resource path; use raw:<verb>")
        a.non_resource_path = res.id
        return a
    if res.type == "namespace":
        if not _ok(NAMESPACE_RE, res.id):
            raise ValueError(f"namespace {go_quote(res.id)} is not a valid namespace name")
        a.namespace = res.id
    elif res.type == "cluster":
        if res.id != "":
            raise ValueError("cluster resources take no id: use cluster?resource=nodes")
    else:
        raise ValueError(f"resource type {go_quote(res.type)}; use namespace:<ns>, cluster or nonresource:<path>")

    req_res, req_group, req_sub = "", "", ""
    q = res.q("resource")
    if q != "":
        req_res, req_group, req_sub = split_resource(q)
    s = res.q("subresource")
    if s != "":
        if not _ok(RESOURCE_RE, s):
            raise ValueError(f"subresource {go_quote(s)} is not valid")
        req_sub = s
    n = res.q("name")
    if n != "":
        if not _ok(NAME_RE, n):
            raise ValueError(f"name {go_quote(n)} is not a valid object name")
        a.name = n
    if res.q("namespace") != "":
        raise ValueError("put the namespace in the resource id: namespace:<ns>")

    if act_res != "" and req_res != "" and (act_res != req_res or act_group != req_group):
        raise ValueError(f"action {action_name} targets {join_res(act_res, act_group)} but the request names {join_res(req_res, req_group)}")
    if act_res != "":
        a.resource, a.group = act_res, act_group
    elif req_res != "":
        a.resource, a.group = req_res, req_group
    elif action_name == "impersonate":
        a.resource = "users"
    elif action_name.startswith("raw:") and ":" not in action_name[4:]:
        raise ValueError(f"action {action_name} is the non-resource form; use it with nonresource:<path>, or add ?resource=<plural>[.<group>] to the request")
    else:
        raise ValueError(f"action {action_name} needs a resource: add ?resource=<plural>[.<group>] to the request")
    if act_sub != "" and req_sub != "" and act_sub != req_sub:
        raise ValueError(f"action {action_name} targets subresource {act_sub} but the request names {req_sub}")
    a.subresource = act_sub if act_sub != "" else req_sub
    return a


def join_res(res: str, group: str) -> str:
    if group == "":
        return res
    return res + "." + group
